File: go/clock/clock.go
package clock

import "fmt"

type Clock struct {
    Hour int
    Minute int
}

func New(h, m int) Clock {
    for m >= 60 {
        m -= 60
        h += 1
    }
	for m < 0 {
        m += 60
        h -= 1
    }
	h = ((h % 24) + 24) % 24
    clock := Clock{Hour: h, Minute: m}
	return clock
}

func (c Clock) Add(m int) Clock {
	return New(c.Hour, c.Minute + m)
}

func (c Clock) Subtract(m int) Clock {
    return New(c.Hour, c.Minute - m)
}

func (c Clock) String() string {
    return fmt.Sprintf("%02d:%02d", c.Hour, c.Minute)
}

File: elixir/clock/lib/clock.ex
defmodule Clock do
  defstruct hour: 0, minute: 0

  @doc """
  Returns a clock that can be represented as a string:

      iex> Clock.new(8, 9) |> to_string
      "08:09"
  """
  @spec new(integer, integer) :: Clock
  def new(hour, minute) do
    total_minutes = rem(rem(60 * hour + minute, 1440) + 1440, 1440)
    %Clock{hour: rem(div(total_minutes, 60), 24), minute: rem(total_minutes, 60)}
  end

  @doc """
  Adds two clock times:

      iex> Clock.new(10, 0) |> Clock.add(3) |> to_string
      "10:03"
  """
  @spec add(Clock, integer) :: Clock
  def add(%Clock{hour: hour, minute: minute}, add_minute), do: new(hour, minute + add_minute)

  defimpl String.Chars, for: Clock do
    def to_string(clock), do: "#{pad_time(clock.hour)}:#{pad_time(clock.minute)}"

    defp pad_time(time) do
      time
      |> Integer.to_string
      |> String.pad_leading(2, "0")
    end
  end
end
